analyzer/consensus/consensus.go

```diff
@@ -648,6 +653,87 @@ func (m *processor) queueRegistryEventInserts(batch *storage.QueryBatch, data *r
 	return nil
 }
 
+func (m *processor) queueRootHashMessageUpserts(batch *storage.QueryBatch, data *rootHashData) error {
+	finalized := map[coreCommon.Namespace]uint64{}
+	var roothashMessageEvents []nodeapi.Event
+	for _, event := range data.Events {
+		switch {
+		case event.RoothashMisc != nil:
+			switch event.Type { //nolint:gocritic,exhaustive // singleCaseSwitch, no special handling for other types
+			case apiTypes.ConsensusEventTypeRoothashFinalized:
+				finalized[event.RoothashMisc.RuntimeID] = *event.RoothashMisc.Round
+			}
+		case event.RoothashExecutorCommitted != nil:
+			runtime := RuntimeFromID(event.RoothashExecutorCommitted.RuntimeID, m.network)
+			if runtime == nil {
+				break
+			}
+			round := event.RoothashExecutorCommitted.Round
+			for i, message := range event.RoothashExecutorCommitted.Messages {
+				messageData := extractMessageData(m.logger.With(
+					"height", data.Height,
+					"runtime", runtime,
+					"round", round,
+					"message_index", i,
+				), message)
+				var relatedAddresses []apiTypes.Address
+				for addr := range messageData.relatedAddresses {
+					relatedAddresses = append(relatedAddresses, addr)
+				}
+				batch.Queue(queries.ConsensusRoothashMessageScheduleUpsert,
+					runtime,
+					round,
+					i,
+					messageData.messageType,
+					messageData.body,
+					relatedAddresses,
+				)
+			}
+		case event.RoothashMessage != nil:
+			// Save these for after we collect all roothash finalized events.
+			roothashMessageEvents = append(roothashMessageEvents, event)
+		}
+	}
+	for _, event := range roothashMessageEvents {
```

Comment:
Can we structure this with some comments? E.g. all the way at the top:
```
1) Extract roothash messages.
```

and then here:
```

2) Extract message results ...
2a) ... from MessageEvents (used in Cobalt).
```

and lower down:
```
2b) ... from LastRoundResults (used in Damask and later).
```

Reply:
k let me see about that

Reply:
adding